---
title: "Stimulated PBMC differential expression analysis"
author: "Delaram PB"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  pdf_document: 
    keep_tex: yes
    number_sections: yes
    toc: yes
  html_document: 
    keep_tex: yes
    toc: yes
  html_notebook: default
  word_document: default
abstract: " "
header-includes: \usepackage{xcolor, colortbl, rotating, graphicx, caption, subcaption}
---

<!--comment
Loading libraries
Setting work directories
-->


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo=FALSE, message = FALSE, warning = FALSE}

source("~/scLMM/LMM-scRNAseq/lmmfit.R")
source("~/scLMM/LMM-scRNAseq/lmmfitSS.R")
source("~/scLMM/LMM-scRNAseq/lmmtest.R")
source("~/scLMM/LMM-scRNAseq/qqpvalue.R")
library(Matrix)
library(MASS)
library(maditr)
library(data.table)
library(Seurat)
library(nebula)
library('glmGamPoi')

```

# Stimulated PBMC data

The stimulated  PBMC data is defined by clusters of the various cell types within the blood in two conditions of control and stimulated. We have 8 individuals ("indiv") from which the cells were derived, a subset (~half) of each individual's sample has been exposed to IFN stimulation. The question is what are the differential expressed genes in the PBMC population between control and stim (with correcting for the patient information).

```{r}
##### PBMC dataset #####
#pbmc <- readRDS('LMM-scRNAseq/Data/PBMC_Lupus_Kang8vs8_data_norm.rds')
pbmc <- readRDS("~/scLMM/LMM-scRNAseq//Data/PBMC_Lupus_Kang8vs8_data_counts.rds")
dim(pbmc)
pbmc = pbmc[rowSums(pbmc)!=0,colSums(pbmc)!=0]
dim(pbmc)
Y_pbmc = GetAssayData(pbmc)
nGenes <- colSums(Y_pbmc)
Ylog_pbmc = log2(t(Y_pbmc) + 1)
Ylog_pbmc <- as.matrix(Ylog_pbmc)
```


```{r}
##design matrix for random effects
pbmc$cell[is.na(pbmc$cell)] = 'unknown'
table(pbmc$cell)
#sum(table(pbmc$cell)) == ncol(pbmc)
Z_pbmc = data.frame(ind=pbmc$ind)
Z_pbmc = model.matrix(~0+ind, Z_pbmc) 
##dimemsion of random effects
d_pbmc <- ncol(Z_pbmc)
```



## Model I: $~ nGenes + cell + stim + (1 | individual)$
Use individual as random effect to consider correlation within the patient. 

```{r}
##design matrix for fixed effects
X_pbmc = data.frame(stim=pbmc$stim,cell=pbmc$cell)
X_pbmc = model.matrix(~ 0 + cell + stim , X_pbmc)
X_pbmc <- cbind(X_pbmc, log_nGenes = log(nGenes))

dim(X_pbmc)
dim(Z_pbmc)
dim(Ylog_pbmc)
```

Fitting model I
```{r, warning = FALSE}
##############################
## **Fit LMM by lmmfit.**
#########################
SEED <- 57774
set.seed(SEED)
maxIter <- 200
t1 <- Sys.time()
fit <- NULL
fit <- lmmfit(Y = Ylog_pbmc, X = X_pbmc, Z = Z_pbmc, 
              d = d_pbmc, max.iter = maxIter, epsilon = 1e-5)
t2 <- Sys.time()
```


```{r}
difftime(t2, t1) 

##number of convergences
sum(fit$niter < maxIter)
##number of non-convergences at epsilon = 1e-5
sum(fit$niter == maxIter)
##at epsilon = 1e-2
sum(apply(abs(fit$dlogL) > 1e-2, 2, any))

##variance components of random effects
k <- 1
range(fit$theta[k, ])
##p-values for hypothesis tests of variance components:
##H0: theta <=0 vs H1: theta > 0
p <- pnorm(fit$theta[k, ]/fit$se[k, ], lower.tail = F)
sum(p <= 0.05)
range(p)

```


Testing the output
```{r, warning = FALSE}
test <- lmmtest(fit)
```


```{r}
test_stim <- test[, grep("stim", colnames(test))]
test_stim <- as.data.frame(test_stim)
test_stim$stim_FDR <- p.adjust(test_stim$stimstim_pvalue, method = "BH") 

test_stim <- test_stim[order(test_stim$stim_FDR),]
sum(test_stim$stim_FDR <= 0.05, na.rm = T)

pv <- test_stim[, grep("pvalue", colnames(test_stim))]
par(mfrow = c(2,1), mar = c(4.5, 4.5, 1.1, 2.1))
qqpvalue(pv, col = "blue", cex = 0.6)
hist(pv, xlab = "Model I p-values for stim", col = "green", cex = 0.6, main=NA)

#fileName <- "Data/model1_test.csv"
#write.csv(cbind(Gene = rownames(test), test), file = fileName, row.names = F)

test_cell <- test[, grep("cell", colnames(test))]
test_cell <- as.data.frame(test_cell)
#test_stim$stim_FDR <- p.adjust(test_cell$stimstim_pvalue, method = "BH") 
#test_stim <- test_stim[order(test_stim$stim_FDR),]
#sum(test_stim$stim_FDR <= 0.05, na.rm = T)

pv <- test_cell[, grep("pvalue", colnames(test_cell))]


par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1))
#par(mfrow = c(2,5), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "", colnames(pv)[i])
  qqpvalue(pv[,i], col = "blue", cex = 0.6, main = main, cex.main = 0.8) }

#dev.off()
par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "", colnames(pv)[i])
  hist(pv[,i], xlab = "p-values", col = "green", cex = 0.6, main = main, cex.main = 0.8) }

```

### evaluating the number of significant DE genes which had a signinficat variance component
```{r}
test_stim <- test[, grep("stim", colnames(test))]
test_stim <- as.data.frame(test_stim)
test_stim$stim_FDR <- p.adjust(test_stim$stimstim_pvalue, method = "BH") 
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(test_stim$stim_FDR < 0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(p<0.05,na.rm = T))

```


### Comparison with NEBULA
## NEBULA I: $~ cell + stim + (1 | individual)$
```{r,warning = FALSE}
##design matrix for fixed effects
X_pbmc = data.frame(stim=pbmc$stim,cell=pbmc$cell)
X_pbmc = model.matrix(~ 1 + cell + stim, X_pbmc)
head(X_pbmc)
num_cells_to_include = ncol(Y_pbmc)


data_g = group_cell(count= as.matrix(Y_pbmc),id=pbmc$ind,pred=X_pbmc)
# t1 <- Sys.time()
# re = nebula(data_g$count,data_g$id,pred=data_g$pred)
#Remove  9935  genes having low expression.
#Analyzing  8955  genes with  8  subjects and  29065  cells.
# t2 <- Sys.time()
# [1] "nebula model I run times:12.050636891524"
# print(difftime(t2, t1))
re <- readRDS('~/scLMM/LMM-scRNAseq/NEBULA_modelI_stimPBMC.rds')
```


Testing model I output
```{r, warning = FALSE}
test <- lmmtest(fit)
```



```{r}
##nebula outputs
##summary (statistics): 
##The estimated coefficient, standard error and p-value for each predictor.
negbn = re
#str(negbn)
st <- negbn$summary
rownames(st) <- st$gene
#head(st)

##The genes which the LMM fit wasn't convergent.
table(fit$niter)
indexNotconverge <- (fit$niter >= 50)
sum(indexNotconverge)

##p-values
##NEBULA
pnb <- as.matrix(st[, grep("p_", colnames(st))])
#head(pnb)

##LMM
plmm <- test[, grep("_p", colnames(test))]
plmm <- plmm[, -grep("log_", colnames(plmm))]
#head(plmm)

i <- (!indexNotconverge)
i_neb = i[row.names(plmm) %in% row.names(pnb)]

par(mfrow = c(2, 1), mar = c(4.1,4.1,1.1,1.1))
hist(plmm[i, -1], xlab = "LMM p-values", main = NA)
hist(pnb[i_neb, -1], xlab = "NEBULA pvalue", main = NA)

##QQ-plot
par(mfrow = c(2, 1), mar = c(4.1,4.1,1.1,1.1))
qqpvalue(plmm[i, -1], main = "LMM p-values", cex.lab = 0.8, cex.main = 0.8)
qqpvalue(pnb[i_neb, -1], main = "NEBULA pvalue", cex.lab = 0.8, cex.main = 0.8)

##number of significant DE genes
##LMM
sum(plmm[i, -1] <= 0.05/sum(i), na.rm = T)

##NEBULA
sum(pnb[i_neb, -1] <= 0.05/sum(i_neb))


#####
##coefficients
bnb <- as.matrix(st[, grep("logFC", colnames(st))])
#head(bnb)
#dim(bnb)

blmm <- test[, grep("_t", colnames(test))]
blmm <- blmm[, -grep("log_", colnames(blmm))]
#head(blmm)
#dim(blmm)
blmm <- blmm[row.names(blmm) %in% row.names(bnb),]
par(mfrow = c(3,3), mar = c(4.1,4.1,1.1,1.1))
for (j in 2:ncol(blmm)) {
	plot(bnb[i_neb, j], blmm[i_neb, j], ylab = "LMM coef", xlab = "NEBULA coef")
  mtext(gsub("_.*", "", colnames(plmm)[j]), line = -1, cex = 0.8)
}

```


## Model II: $~ cell + stim:cell + (1 | individual)$
Use individual as random effect to consider correlation within the patient. 

```{r}
##design matrix for fixed effects
X_pbmc = data.frame(stim=pbmc$stim,cell=pbmc$cell)
X_pbmc = model.matrix(~ 0 + cell + stim:cell, X_pbmc)

dim(X_pbmc)
dim(Z_pbmc)
dim(Ylog_pbmc)
```


Fitting model II
```{r, warning = FALSE}
##############################
## **Fit LMM by lmmfit.**
#########################
SEED <- 57774
set.seed(SEED)
maxIter <- 200
t1 <- Sys.time()
fit <- NULL
fit <- lmmfit(Y = Ylog_pbmc, X = X_pbmc, Z = Z_pbmc, 
              d = d_pbmc, max.iter = maxIter, epsilon = 1e-5)
t2 <- Sys.time()
```


```{r}
difftime(t2, t1) 

##number of convergences
sum(fit$niter < maxIter)
##number of non-convergences at epsilon = 1e-5
sum(fit$niter == maxIter)
##at epsilon = 1e-2
sum(apply(abs(fit$dlogL) > 1e-2, 2, any))

##variance components of random effects
k <- 1
range(fit$theta[k, ])
##p-values for hypothesis tests of variance components:
##H0: theta <=0 vs H1: theta > 0
p <- pnorm(fit$theta[k, ]/fit$se[k, ], lower.tail = F)
sum(p <= 0.05)
range(p)

```

Testing model II output
```{r, warning = FALSE}
test_org <- lmmtest(fit)
```


```{r}
test <- test_org[, grep("stim", colnames(test_org))]
test <- as.data.frame(test)
test$stim_FDR <- p.adjust(test$`cellCD8 T cells:stimstim_pvalue`, method = "BH") 

test <- test[order(test$stim_FDR),]
sum(test$stim_FDR <= 0.05, na.rm = T)

pv <- test[, grep("pvalue", colnames(test))]
par(mfrow = c(2,1), mar = c(4.5, 4.5, 1.1, 2.1))
qqpvalue(pv, col = "blue", cex = 0.6)
hist(pv$`cellCD8 T cells:stimstim_pvalue`, xlab = "Model II p-values for cellCD8 T cells:stim", col = "green", cex = 0.6, main=NA)

#fileName <- "Data/model1_test.csv"
#write.csv(cbind(Gene = rownames(test), test), file = fileName, row.names = F)

par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1))
#par(mfrow = c(2,5), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "-wildtype", colnames(pv)[i])
  qqpvalue(pv[,i], col = "blue", cex = 0.6, main = main, cex.main = 0.8) }

#dev.off()
par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "-wildtype", colnames(pv)[i])
  hist(pv[,i], xlab = "p-values", col = "green", cex = 0.6, main = main, cex.main = 0.8) }

```


### evaluating the number of significant DE genes which had a signinficat variance component
```{r}
test_stim <- test_org[, grep("stim", colnames(test_org))]
test_stim <- as.data.frame(test_stim)
test_stim$stim_FDR <- p.adjust(test_stim$`cellCD8 T cells:stimstim_t`, method = "BH") 
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(test_stim$stim_FDR < 0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(p<0.05,na.rm = T))

```

## Model III: $~ nGenes + cell + stim:cell + (1 | individual)$
Use individual as random effect to consider correlation within the patient. 

```{r}
##design matrix for fixed effects
X_pbmc = data.frame(stim=pbmc$stim,cell=pbmc$cell)
X_pbmc = model.matrix(~ 0 + cell + stim:cell, X_pbmc)
X_pbmc <- cbind(X_pbmc, log_nGenes = log(nGenes))

dim(X_pbmc)
dim(Z_pbmc)
dim(Ylog_pbmc)
```


Fitting model III
```{r, warning = FALSE}
##############################
## **Fit LMM by lmmfit.**
#########################
SEED <- 57774
set.seed(SEED)
maxIter <- 200
t1 <- Sys.time()
fit <- NULL
fit <- lmmfit(Y = Ylog_pbmc, X = X_pbmc, Z = Z_pbmc, 
              d = d_pbmc, max.iter = maxIter, epsilon = 1e-5)
t2 <- Sys.time()
```


```{r}
difftime(t2, t1) 

##number of convergences
sum(fit$niter < maxIter)
##number of non-convergences at epsilon = 1e-5
sum(fit$niter == maxIter)
##at epsilon = 1e-2
sum(apply(abs(fit$dlogL) > 1e-2, 2, any))

##variance components of random effects
k <- 1
range(fit$theta[k, ])
##p-values for hypothesis tests of variance components:
##H0: theta <=0 vs H1: theta > 0
p <- pnorm(fit$theta[k, ]/fit$se[k, ], lower.tail = F)
sum(p <= 0.05)
range(p)

```

Testing model III output
```{r, warning = FALSE}
test_org <- lmmtest(fit)
```


```{r}
test <- test_org[, grep("stim", colnames(test_org))]
test <- as.data.frame(test)
test$stim_FDR <- p.adjust(test$`cellCD8 T cells:stimstim_pvalue`, method = "BH") 

test <- test[order(test$stim_FDR),]
sum(test$stim_FDR <= 0.05, na.rm = T)

pv <- test[, grep("pvalue", colnames(test))]
par(mfrow = c(2,1), mar = c(4.5, 4.5, 1.1, 2.1))
qqpvalue(pv, col = "blue", cex = 0.6)
hist(pv$`cellCD8 T cells:stimstim_pvalue`, xlab = "Model III p-values for cellCD8 T cells:stim", col = "green", cex = 0.6, main=NA)

#fileName <- "Data/model1_test.csv"
#write.csv(cbind(Gene = rownames(test), test), file = fileName, row.names = F)

par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1))
#par(mfrow = c(2,5), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "-wildtype", colnames(pv)[i])
  qqpvalue(pv[,i], col = "blue", cex = 0.6, main = main, cex.main = 0.8) 
  }

#dev.off()
par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "-wildtype", colnames(pv)[i])
  hist(pv[,i], xlab = "p-values", col = "green", cex = 0.6, main = main, cex.main = 0.8) 
  }

```



### evaluating the number of significant DE genes which had a signinficat variance component
```{r}
test_stim <- test_org[, grep("stim", colnames(test_org))]
test_stim <- as.data.frame(test_stim)
test_stim$stim_FDR <- p.adjust(test_stim$`cellCD8 T cells:stimstim_t`, method = "BH") 
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(test_stim$stim_FDR < 0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(p<0.05,na.rm = T))

```


### Comparison with NEBULA
## NEBULA II: $~ cell + stim:cell + (1 | individual)$
```{r,warning = FALSE}
##design matrix for fixed effects
X_pbmc = data.frame(stim=pbmc$stim,cell=pbmc$cell)
X_pbmc = model.matrix(~ 1 + cell + stim:cell, X_pbmc)
head(X_pbmc)
num_cells_to_include = ncol(Y_pbmc)

data_g = group_cell(count= as.matrix(Y_pbmc),id=pbmc$ind,pred=X_pbmc)
#t1 <- Sys.time()
#re = nebula(data_g$count,data_g$id,pred=data_g$pred)
#Remove  9935  genes having low expression.
#Analyzing  8955  genes with  8  subjects and  29065  cells.
#t2 <- Sys.time()
#difftime(t2, t1)
#Time difference of 23.57783 mins
re <- readRDS('~/scLMM/LMM-scRNAseq/NEBULA_modelII_stimPBMC.rds')

```



Testing model III output
```{r, warning = FALSE}
test <- lmmtest(fit)
```


```{r}
##nebula outputs
##summary (statistics): 
##The estimated coefficient, standard error and p-value for each predictor.
negbn = re
#str(negbn)
st <- negbn$summary
rownames(st) <- st$gene
#head(st)

##The genes which the LMM fit wasn't convergent.
table(fit$niter)
indexNotconverge <- (fit$niter >= 50)
sum(indexNotconverge)

##p-values
##NEBULA
pnb <- as.matrix(st[, grep("p_", colnames(st))])
#head(pnb)

##LMM
plmm <- test[, grep("_p", colnames(test))]
#plmm <- plmm[, -grep("log_", colnames(plmm))]
#head(plmm)

i <- (!indexNotconverge)
i_neb = i[row.names(plmm) %in% row.names(pnb)]

par(mfrow = c(2, 1), mar = c(4.1,4.1,1.1,1.1))
hist(plmm[i, -1], xlab = "LMM p-values", main = NA)
hist(pnb[i_neb, -1], xlab = "NEBULA pvalue", main = NA)

##QQ-plot
par(mfrow = c(2, 1), mar = c(4.1,4.1,1.1,1.1))
qqpvalue(plmm[i, -1], main = "LMM p-values", cex.lab = 0.8, cex.main = 0.8)
qqpvalue(pnb[i_neb, -1], main = "NEBULA pvalue", cex.lab = 0.8, cex.main = 0.8)

##number of significant DE genes
##LMM
sum(plmm[i, -1] <= 0.05/sum(i), na.rm = T)

##NEBULA
sum(pnb[i_neb, -1] <= 0.05/sum(i_neb))


#####
##coefficients
bnb <- as.matrix(st[, grep("logFC", colnames(st))])
#head(bnb)
#dim(bnb)

blmm <- test[, grep("_t", colnames(test))]
blmm <- blmm[, -grep("log_", colnames(blmm))]
#head(blmm)
#dim(blmm)
blmm <- blmm[row.names(blmm) %in% row.names(bnb),]
par(mfrow = c(3,3), mar = c(4.1,4.1,1.1,1.1))
for (j in 2:ncol(blmm)) {
	plot(bnb[i_neb, j], blmm[i_neb, j], ylab = "LMM coef", xlab = "NEBULA coef")
  mtext(gsub("_.*", "", colnames(plmm)[j]), line = -1, cex = 0.8)
}

```



```{r}
##### PBMC dataset #####
#pbmc <- readRDS('LMM-scRNAseq/Data/PBMC_Lupus_Kang8vs8_data_norm.rds')
pbmc <- readRDS("~/scLMM/LMM-scRNAseq//Data/PBMC_Lupus_Kang8vs8_data_counts.rds")
dim(pbmc)
pbmc = pbmc[rowSums(pbmc)!=0,colSums(pbmc)!=0]
dim(pbmc)
pbmc = Seurat::SCTransform(pbmc, 'originalexp')


Y_pbmc = GetAssayData(pbmc)
Y_pbmc = Y_pbmc[rowSums(Y_pbmc)!=0,colSums(Y_pbmc)!=0]

nGenes <- colSums(Y_pbmc)
Ylog_pbmc = t(Y_pbmc)
#Ylog_pbmc = log2(t(Y_pbmc) + 1)
Ylog_pbmc <- as.matrix(Ylog_pbmc)
```


```{r}
##design matrix for random effects
pbmc$cell[is.na(pbmc$cell)] = 'unknown'
table(pbmc$cell)
#sum(table(pbmc$cell)) == ncol(pbmc)
Z_pbmc = data.frame(ind=pbmc$ind)
Z_pbmc = model.matrix(~0+ind, Z_pbmc) 
##dimemsion of random effects
d_pbmc <- ncol(Z_pbmc)
```



## Model IV: $~ nGenes + cell + cell:stim + (1 | individual)$
Use individual as random effect to consider correlation within the patient. 

```{r}
##design matrix for fixed effects
X_pbmc = data.frame(stim=pbmc$stim,cell=pbmc$cell)
X_pbmc = model.matrix(~ 0 + cell + cell:stim , X_pbmc)
X_pbmc <- cbind(X_pbmc, log_nGenes = log(nGenes))

dim(X_pbmc)
dim(Z_pbmc)
dim(Ylog_pbmc)
```

Fitting model IV
```{r, warning = FALSE}
##############################
## **Fit LMM by lmmfit.**
#########################
SEED <- 57774
set.seed(SEED)
maxIter <- 200
t1 <- Sys.time()
fit <- NULL
fit <- lmmfit(Y = Ylog_pbmc, X = X_pbmc, Z = Z_pbmc, 
              d = d_pbmc, max.iter = maxIter, epsilon = 1e-5)
t2 <- Sys.time()
```


```{r}
difftime(t2, t1) 

##number of convergences
sum(fit$niter < maxIter)
##number of non-convergences at epsilon = 1e-5
sum(fit$niter == maxIter)
##at epsilon = 1e-2
sum(apply(abs(fit$dlogL) > 1e-2, 2, any))

##variance components of random effects
k <- 1
range(fit$theta[k, ])
##p-values for hypothesis tests of variance components:
##H0: theta <=0 vs H1: theta > 0
p <- pnorm(fit$theta[k, ]/fit$se[k, ], lower.tail = F)
sum(p <= 0.05)
range(p)

```



Testing model IV output
```{r, warning = FALSE}
test_org <- lmmtest(fit)
```


```{r}
test <- test_org[, grep("stim", colnames(test_org))]
test <- as.data.frame(test)
test$stim_FDR <- p.adjust(test$`cellCD8 T cells:stimstim_pvalue`, method = "BH") 

test <- test[order(test$stim_FDR),]
sum(test$stim_FDR <= 0.05, na.rm = T)

pv <- test[, grep("pvalue", colnames(test))]
par(mfrow = c(2,1), mar = c(4.5, 4.5, 1.1, 2.1))
qqpvalue(pv, col = "blue", cex = 0.6)
hist(pv$`cellCD8 T cells:stimstim_pvalue`, xlab = "Model IV p-values for cellCD8 T cells:stim", col = "green", cex = 0.6, main=NA)

#fileName <- "Data/model1_test.csv"
#write.csv(cbind(Gene = rownames(test), test), file = fileName, row.names = F)

par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1))
#par(mfrow = c(2,5), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "-wildtype", colnames(pv)[i])
  qqpvalue(pv[,i], col = "blue", cex = 0.6, main = main, cex.main = 0.8) 
  }

#dev.off()
par(mfrow = c(3,2), mar = c(5.1, 4.1, 3.1, 1.1)) 
for (i in 1:ncol(pv)) {
  main <- gsub("_pvalue", "-wildtype", colnames(pv)[i])
  hist(pv[,i], xlab = "p-values", col = "green", cex = 0.6, main = main, cex.main = 0.8) 
  }

```


### evaluating the number of significant DE genes which had a signinficat variance component
```{r}
test_stim <- test_org[, grep("stim", colnames(test_org))]
test_stim <- as.data.frame(test_stim)
test_stim$stim_FDR <- p.adjust(test_stim$`cellCD8 T cells:stimstim_t`, method = "BH") 
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(test_stim$stim_FDR < 0.05,na.rm = T))
print(sum(test_stim$stim_FDR < 0.05 & p<0.05,na.rm = T)/sum(p<0.05,na.rm = T))

```

